src/k2/module/k23si/Module.cpp

```diff
@@ -371,15 +373,20 @@ seastar::future<> K23SIPartitionModule::start() {
 
     _registerMetrics();
 
-    _cpo.init(_config.cpoEndpoint());
+    K2LOG_I(log::skvsvr, "init cpo with {}", _cpoEndpoint);
+    _cpo.init(_cpoEndpoint);
     if (_cmeta.retentionPeriod < _config.minimumRetentionPeriod()) {
         K2LOG_W(log::skvsvr,
             "Requested retention({}) is lower than minimum({}). Extending retention to minimum",
             _cmeta.retentionPeriod, _config.minimumRetentionPeriod());
         _cmeta.retentionPeriod = _config.minimumRetentionPeriod();
     }
 
-    return _tsoClient.getTimestamp()
+    return AppBase().getDist<tso::TSOClient>().invoke_on(seastar::this_shard_id(), &tso::TSOClient::bootstrap,
```

Comment:
_tsoClient is already the local instance - I think you can just call _tsoClient.bootstrap(_cpoEndpoint) here.